### PR TITLE
Lookup a layer in reverse order (enhancement)

### PR DESCRIPTION
**Pull request by gx740**
In many cases the lookup in reverse order in packet is faster than in the direct one.

This PR adds such feature and improvements in a number of modules.

### CODE REVIEW

**Comment by gx740** [comment]
For example: when a packet is being passed into IP reassembly logic the last layer, usually, is a IPv4/v6 so the searching has two operations: `dynamic_cast` and comparing to `NULL`.

**Comment by gx740** [comment]
I was wrong. An example above is incorrect. Currently the user is forced to make parsing until the layer he/she interested in. So the latest layer will not be the IPv4/v6.

File: Tests/Packet++Test/main.cpp [hunk 1]
```diff
@@ -7561,6 +7561,32 @@ PTF_TEST_CASE(EthDot3LayerCreateEditTest)
 
 
 
+PTF_TEST_CASE(PacketLayerLookupTest)
```
Comment (seladb) [inline Tests/Packet++Test/main.cpp]: I'd suggest to add a few tests:

- Packet that has a few layers of the same type (you can use `Vxlan1.dat`)
- Try to get prev layer of the first layer (result should be `NULL`)
- Try `getLayerOfType<...>(true)` with a layer type that doesn't exist in the packet

Comment (gx740) [reply Tests/Packet++Test/main.cpp]: I will do it today

Comment (gx740) [reply Tests/Packet++Test/main.cpp]: Done

**Pull request merged**